### PR TITLE
feat(connector-iroha): add dynamic request params

### PR DESCRIPTION
**Pull request by yashrajdesai**
Fixes: [#1253](https://github.com/hyperledger/cactus/issues/1253)

Signed-off-by: yashrajdesai <yashrajdesai30@gmail.com>

### CODE REVIEW

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:325]: This is optional, not required:
https://iroha.readthedocs.io/en/develop/develop/api/commands.html#compare-and-set-account-detail

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:406]: "number" can be float/double, but here I suggest type "integer"

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:487]: The command https://iroha.readthedocs.io/en/develop/develop/api/queries.html#get-account-detail has few params optional

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:528]: this is optional, according to: https://iroha.readthedocs.io/en/develop/develop/api/queries.html#get-account-assets

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:692]:
```suggestion
            "type": "integer"
```

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:705]: this is optional.

Also there are more optional params:
https://iroha.readthedocs.io/en/develop/develop/api/queries.html#result-pagination
Look to the picture:
obraz

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:726]: Comments as for: `GetAccountTransactionsRequestParameters`

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:712]: "integer"

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:751]: integer

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:797]: There is also optional TLS certificate: https://iroha.readthedocs.io/en/develop/develop/api/commands.html#add-peer

File: packages/cactus-plugin-ledger-connector-iroha/src/main/typescript/plugin-ledger-connector-iroha.ts
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/typescript/plugin-ledger-connector-iroha.ts:309]:
```suggestion
          const response = await commands
```

**Review by baziorek — CHANGES_REQUESTED** [review]
Few corrections according to iroha's api required

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:274]: One more suggestion - order of references.
I see order of entries:
1. CreateAccountRequestParameters
2. SetAccountDetailRequestParameters
3. CompareAndSetAccountDetailRequestParameters
4. CreateAssetRequestParameters
5. CreateDomainRequestParameters
6. SetAccountQuorumRequestParameters
7. AddAssetQuantityRequestParameters
8. SubtractAssetQuantityRequestParameters
9. TransferAssetRequestParameters
10. GetSignatoriesRequestParameters
11. GetAccountRequestParameters
12. GetAccountDetailRequestParameters
13. GetAssetInfoRequestParameters
14. GetAccountAssetsRequestParameters
15. AddSignatoryRequestParameters
16. RemoveSignatoryRequestParameters
17. CreateRoleRequestParameters
18. AppendRoleRequestParameters
19. DetachRoleRequestParameters
20. GetRolePermissionsRequestParameters
21. GrantPermissionRequestParameters
22. RevokePermissionRequestParameters
23. GetTransactionsRequestParameters
24. GetPendingTransactionsRequestParameters
25. GetAccountTransactionsRequestParameters
26. GetAccountAssetTransactionsRequestParameters
27. GetBlockRequestParameters
28. CallEngineRequestParameters
29. GetEngineReceiptsRequestParameters
30. AddPeerRequestParameters
31. RemovePeerRequestParameters
But it looks like random order. I suggest: first commands then queries. Then if there is no specific order in Cactus's style I suggest to use the same order as in Iroha: https://iroha.readthedocs.io/en/develop/develop/api.html - it is alphabetically

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (yashrajdesai) [reply packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:705]: So shall I add firstTxTime, lastTxTime, firstTxHeight, lastTxHeight as well in the GetAccountTransactionsRequestParameters?

Comment (baziorek) [reply packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:705]: > So shall I add firstTxTime, lastTxTime, firstTxHeight, lastTxHeight as well in the GetAccountTransactionsRequestParameters?

I think it would be good to add them as optional params (of course if Cactus can handle that params)

**Comment by yashrajdesai** [comment]
@baziorek I have included all the above suggested changes in the recent commit, please review.

**Comment by baziorek** [comment]
> @baziorek I have included all the above suggested changes in the recent commit, please review.

It is great! I'll take a look tomorrow. But please change status of resolved comments from me into **resolved**.

BTW. I think it is not good to remove @petermetz from reviewers list :D.

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [reply packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:274]: - [x] I see that it is sorted after update.

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:290]: amount can be double: 200.02, so I suggest to restore:
```suggestion
            "type": "number"
```

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:310]: AddPeer should contain one more param:
https://iroha.readthedocs.io/en/develop/develop/api/commands.html#add-peer
`syncing_peer`

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:357]: This looks strange for me. Is everything OK with the code? I'm not expert with OpenAPI, but other properties don't contain that fields.

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:607]: According to: https://iroha.readthedocs.io/en/develop/develop/api/commands.html#compare-and-set-account-detail also `check_empty` is required

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:689]: I suggest to add also optional params
- `first_tx_time`
- `last_tx_time`
In [documentation](https://iroha.readthedocs.io/en/develop/develop/api/queries.html#result-pagination) there are more optional params - but they are useless in the command (pending transaction is not committed, so it is not in any block, so filtering to specific blocks is not useful)

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:715]: I'm thinking about those types. According to iroha documentation they are timestamps: https://iroha.readthedocs.io/en/develop/develop/api/queries.html#result-pagination (Timestamps with milliseconds). So I don't know if it is good to use just `string`, or maybe `integer` or `date-time`.

We will see what @petermetz would decide.

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [reply packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:731]: - [x] I see it is resolved after code update

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:828]: It can contains not only numbers, but also letters, so I suggest:
```suggestion
            "type": "string"
```

**Review by baziorek — CHANGES_REQUESTED** [review]
Few more corrections in OpenAPI.json required

**Comment by yashrajdesai** [comment]
> > @baziorek I have included all the above suggested changes in the recent commit, please review.
> 
> It is great! I'll take a look tomorrow. But please change status of resolved comments from me into **resolved**.
> 
> BTW. I think it is not good to remove @petermetz from reviewers list :D.

Actually I had requested a review from you using the re-request feature but when I did that it automatically removed all the reviewers from reviewers section and now I am not able to add any reviewers as I do not have the access to do it.

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (yashrajdesai) [reply packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:310]: The current cactus version doesn't support the syncing_peer field. So can we keep it optional until the cactus is updated with the latest version?

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (yashrajdesai) [reply packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:715]: Sure, I will update it accordingly.

File: packages/cactus-plugin-ledger-connector-iroha/src/main/typescript/plugin-ledger-connector-iroha.ts
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/typescript/plugin-ledger-connector-iroha.ts:845]: I'm not expert, but one thing looks for me suspiciously: what would happen when not all params would be provided? Or in different order?

E.g.:    
obraz

Comment (yashrajdesai) [reply packages/cactus-plugin-ledger-connector-iroha/src/main/typescript/plugin-ledger-connector-iroha.ts:845]: image
Exactly that's the problem which this PR is going to fix as you can see in the else block we accept the request params in any order. The reason for still keeping the hardcoded way of params is backward compatibility. It would be removed in the future releases. (As discussed with @petermetz)

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (petermetz) [reply packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:721]: > Sure, I will update it accordingly.

@yashrajdesai My answer is that whichever works (e.g. results in the correct automatic conversions) should be used. Right now, we don't know which one that would be though because the test case does not use these parameters at all, so the way to go is to add another test case within the existing test file where these parameters get used and then write an assertion to verify that the parameters are working as intended. If they don't, then we can tweak the types of the request properties until they do.

File: packages/cactus-plugin-ledger-connector-iroha/src/main/typescript/plugin-ledger-connector-iroha.ts
Comment (petermetz) [reply packages/cactus-plugin-ledger-connector-iroha/src/main/typescript/plugin-ledger-connector-iroha.ts:845]: @yashrajdesai @baziorek Yup, the inflexibility of the array based approach is what we are fixing here, but in a backward compatible way. We can drop the array based approach with the release of 2.0 and I already asked @yashrajdesai to create an issue in the tracker specifically for that so for now this thread can be resolved once that GH issue has been opened.

**Review by petermetz — CHANGES_REQUESTED** [review]
@yashrajdesai 99% LGTM but please take a look at this and address it: https://github.com/hyperledger/cactus/pull/2148#discussion_r954282680

**Comment by baziorek** [comment]
> @yashrajdesai 99% LGTM but please take a look at this and address it: [#2148 (comment)](https://github.com/hyperledger/cactus/pull/2148#discussion_r954282680)

For me it is also **almost** LGTM. The last comment from me is linked by Peter, so probably after resolving its I'll approve. Probably - because I'd like to review everything again to make sure if nothing escaped my attention:).

File: packages/cactus-plugin-ledger-connector-iroha/src/main/typescript/plugin-ledger-connector-iroha.ts
Comment (yashrajdesai) [reply packages/cactus-plugin-ledger-connector-iroha/src/main/typescript/plugin-ledger-connector-iroha.ts:845]: @petermetz @baziorek Opened issue [#2151](https://github.com/hyperledger/cactus/issues/2151)  for the same.

File: packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json
Comment (baziorek) [inline packages/cactus-plugin-ledger-connector-iroha/src/main/json/openapi.json:792]:
```suggestion
            "type": "integer"
```
https://iroha.readthedocs.io/en/develop/develop/api/queries.html#get-account-detail and rather: https://iroha.readthedocs.io/en/develop/develop/api/queries.html#result-pagination

**Review by baziorek — APPROVED** [review]
Only small changes `"number"` -> `"integer"`.

Rest of things connected with Iroha 1 is OK

**Comment by yashrajdesai** [comment]
> @yashrajdesai 99% LGTM but please take a look at this and address it: [#2148 (comment)](https://github.com/hyperledger/cactus/pull/2148#discussion_r954282680)

I have addressed it [here](https://github.com/yashrajdesai/cactus/blob/dynamic-req-params/packages/cactus-plugin-ledger-connector-iroha/src/test/typescript/integration/run-transaction-endpoint-v1.test.ts#L1014).

**Comment by petermetz** [comment]
> firstTxTime

@yashrajdesai I looked through it, seems legit. Next time please re-request the review once you've made the changes. That way GitHub shows that I have a pending review in my queue (right now this doesn't I only saw this because you told me in the pair prog sessions - which also works but it's more efficient for everybody if we use the review queue otherwise I may forget)

**Review by petermetz — APPROVED** [review]
@yashrajdesai LGTM

**Comment by petermetz** [comment]
@yashrajdesai We need one more maintainer to approve it from either Fujitsu or IBM so please try to find reviewers from: @VRamakrishna @sanvenDev @takeutak @izuru0

**Comment by takeutak** [comment]
@petermetz 
> [petermetz](https://github.com/petermetz) commented [2 days ago](https://github.com/hyperledger/cactus/pull/2148#issuecomment-1276778564)
@yashrajdesai We need one more maintainer to approve it from either Fujitsu or IBM so please try to find reviewers from: @VRamakrishna @sanvenDev @takeutak @izuru0

Sorry for my late replying.  Good jobs for this pull-requests!
I'm asking @izuru0 to review this pull-request.  He will review this pull-request in a few days.  I'm sorry for waiting you due to the time difference (Today Japan is Saturday).

**Review by jagpreetsinghsasan — APPROVED** [review]

**Comment by izuru0** [comment]
LGTM

**Pull request merged**
